feat(feature:account): Migrate account module to CMP

HekmatullahAmin opened the pull request:
Fixes - [Jira-#178](https://mifosforge.jira.com/browse/MM-178)

Didn't create a Jira ticket, click [here](https://mifosforge.jira.com/jira/software/c/projects/MM/issues/) to create new.

Please Add Screenshots If there are any UI changes.

| Before                                     | After                                  |
|--------------------------------------------|----------------------------------------|
|                                            |                                        |


Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Run the static analysis check `./gradlew check` or `ci-prepush.sh` to make sure you didn't break anything

- [ ] If you have multiple commits please combine them into one commit by squashing them.

HekmatullahAmin committed:
refactor(feature:account): Set up required dependencies for the module

HekmatullahAmin committed:
refactor(feature:account): Move file to commonMain and resolve Detekt issues

HekmatullahAmin committed:
feat(feature:account): Migrate account module to CMP

therajanmaurya commented on core/designsystem/src/commonMain/kotlin/org/mifos/mobile/core/designsystem/component/MifosButton.kt:
```diff
@@ -143,6 +143,7 @@ fun MifosOutlinedButton(
  * clickable and will appear disabled to accessibility services.
  * @param content The button content.
  */
+// internal fun MifosTextButton(
```
Remove this

therajanmaurya commented on feature/account/src/commonMain/kotlin/org/mifos/mobile/feature/account/account/screens/AccountsScreen.kt:
```diff
@@ -10,44 +10,44 @@
 
 @Composable
 internal fun AccountsScreen(
+    isNetworkConnected: Boolean,
```
Inject NetworkMonitor directory in AccountsViewModel and observe as AccountsViewModel in here and composable and use

therajanmaurya commented on feature/account/src/commonMain/kotlin/org/mifos/mobile/feature/account/clientAccount/screens/ClientAccountsScreen.kt:
```diff
@@ -23,37 +24,40 @@ import androidx.compose.runtime.saveable.rememberSaveable
 
 @Composable
 internal fun ClientAccountsScreen(
+    isNetworkConnected: Boolean,
```
Inject NetworkMonitor directory in AccountsViewModel and observe as AccountsViewModel in here and composable and use

niyajali commented on feature/account/src/commonMain/kotlin/org/mifos/mobile/feature/account/Color.kt:
```diff
@@ -0,0 +1,31 @@
+
+import androidx.compose.ui.graphics.Color
+
+val White = Color(0xFFFFFFFF)
```
Why, you're adding and using these HardCoded colors, use MaterialColor, never use hardcoded colors to a Composable,
remove all hardcoded color from all composable and use MaterialColor instead

HekmatullahAmin replied:
@niyajali  I completely get your point about avoiding hardcoded colors. Previously, the account module had a colors.xml file, but since Compose Multiplatform can’t access it, I created a Color.kt with the same colors and used it in the code.

If I were to move these colors into the designsystem module and set up light/dark theme configurations, it would mean changing another module, which isn’t within the scope of this PR.

That said, my idea was to keep it as is for now and, when we work on UI colors and typography in the future, we can centralize everything properly. What do you think about this approach?

<img width="765" alt="Screenshot 2025-02-13 at 07 52 37" src="https://github.com/user-attachments/assets/914dcbc2-50b4-4503-bcf3-6bf9c628600c" />

niyajali replied:
Could you tell me where we exactly using these colours, like which type of composable like it's a basic text, Badge, card etc, if so use M3 colors like, territory, secondary, error, etc we've lots of M3 colors, use those no need to use those hardcoded color or no need to move to design system.

niyajali replied:
And we don't need to configure any design system elements like typography, colors for different modes, or for dynamic themes, since we already configured everything on project setup, and if we pass hardcoded value to any composable we've to manually configure/provide colors for different mode, otherwise it will be automatically provided/configured by M3. And it's best practices not to pass hardcoded value to any composable, we could use M3 color and provide alpha for more configuration.

And we could tell more if you provide a screenshot of that screen, so we can see how it looks, and can tell which other Jetpack composables can be used for better usability and less configuration to look the same.

HekmatullahAmin replied:
The colors are primarily used to indicate different account states like Approved, Submitted, Closed, Withdrawn, etc. They’re also applied in the Filter dialog checkboxes to help filter these accounts. I’ve attached a screenshot below for better context.

I totally get your point about following M3 principles and avoiding hardcoded colors. If you’d like, I can go ahead and apply the best-suited colors from the M3 palette based on these states, ensuring we stick to best practices.

<img src="https://github.com/user-attachments/assets/164ec3bf-9d8e-46de-a913-570a9f4e96f0" width="300" height="500"/>
<img src="https://github.com/user-attachments/assets/aab9c50a-370f-4077-87ef-3c4f0847136b" width="300" height="500"/>
<img src="https://github.com/user-attachments/assets/5984ec22-bb64-4294-8b4f-e11f41c29b02" width="300" height="500"/>

niyajali commented:
@HekmatullahAmin And I don't understand what you are trying to do, Your PR title says `setting up required dependencies`, but you're changing the production(UI) codebase, could you please explain it

HekmatullahAmin commented:
> @HekmatullahAmin And I don't understand what you are trying to do, Your PR title says `setting up required dependencies`, but you're changing the production(UI) codebase, could you please explain it

@niyajali Thanks for the feedback! You’re absolutely right, the PR title doesn’t reflect the changes accurately. Initially, I followed a step-by-step migration process: first, I added the required dependencies, then moved files to common, and finally migrated to CMP as needed. Because of that, the title remained the same. I’ll update it now to better reflect the changes.

HekmatullahAmin committed:
refactor: Replace hardcoded colors with MaterialTheme colors and add NetworkMonitor

niyajali commented:
@HekmatullahAmin I've generated those colors using the primary color only. Besides, providing different colors for checkbox is just useless and doesn't look good, use the default color since it contained a label which indicates its value and redesigns the account screen item card using `Card/ListItem`  and add some badges for different states like active or inactive, approved, etc.

niyajali commented on feature/account/src/commonMain/kotlin/org/mifos/mobile/feature/account/Color.kt:
```diff
@@ -0,0 +1,23 @@
+
+import androidx.compose.ui.graphics.Color
+
+val Black = Color(0xFF000000)
```
Remove these colors we no need to provide separate colors for a feature, if we're doing like for every feature then configuring design system and M3 is just useless

niyajali commented on feature/account/src/commonMain/kotlin/org/mifos/mobile/feature/account/viewmodel/AccountsViewModel.kt:
```diff
@@ -256,12 +260,12 @@ class AccountsViewModel @Inject constructor(
             account?.let {
-                it.productName?.lowercase(Locale.ROOT)?.contains(searchTerm) == true ||
-                    it.accountNo?.lowercase(Locale.ROOT)?.contains(searchTerm) == true
+                it.productName?.lowercase()?.contains(searchTerm) == true ||
```
Move all filtering logic to separate UseCase, and use IODispatcher for filtering or sorting items, this ViewModel looks messy to me.

HekmatullahAmin committed:
Wrapped account items in cards, removed hardcoded colors, and applied default checkbox colors

HekmatullahAmin commented:
Hi @niyajali 

I just wanted to let you know about the recent changes I made:

- Account items are now wrapped inside cards for better UI consistency.
- I removed hardcoded colors and replaced them with theme-based styling.
- The checkbox components now use their default colors, making the code cleaner and easier to maintain.

Regarding your review about moving the filtering logic to a separate UseCase and using IODispatcher, I totally agree with you. I also noticed the ViewModel feels a bit messy—for instance, the refresh and setFilterList methods could be simplified, and we could have a generic searchInList function instead of separate methods for searchInSharesList, searchInLoanList, and searchInSavingsList.

When I picked up the task, I actually brought up the idea of cleaning things up and aligning on a pattern during stand-up. Since the Mobile Wallet project uses MVI, I asked if we should consider consistency across projects. But the response was that the primary focus right now is just to complete the migration, and we can revisit the structure and clean-up efforts later.

HekmatullahAmin commented:
@niyajali Based on what I mentioned earlier, I didn’t make changes to the ViewModel because the focus right now is on completing the migration, as discussed in the stand-up. However, if you still feel it’s important to address those changes now, please let me know, and I’ll make the updates accordingly.
Otherwise, if everything looks good, could you please approve and merge it?

niyajali commented:
@HekmatullahAmin can't merge it, as we discussed in slack create a quick pr to setup all accounts modules and setup base module with tab setup, and then send separate PR's for each module
